mqtt/mqtt-bridge/tests/integration.rs

```diff
@@ -128,6 +133,100 @@ async fn send_message_upstream_downstream() {
     local_client.shutdown().await;
 }
 
+/// Scenario:
+///	- Creates 2 brokers and a bridge to connect between the brokers.
+///	- A client connects to local broker and subscribes to receive messages from upstream
+/// - A client connects to remote broker and subscribes to receive messages from downstream
+/// - Client publish message upstream
+///	- Expects to receive messages downstream -> upstream
+/// - Client publish message upstream
+/// - Disconnect bridge and shutdown (simulate process restart)
+/// - Reconnect bridge
+/// - Expects to receive messages downstream -> upstream
+#[tokio::test]
+async fn send_message_upstream_with_crash_is_lossless() {
+    let subs = vec![Direction::Out(TopicRule::new(
+        "temp/#".into(),
+        Some("to".into()),
+        Some("upstream".into()),
+    ))];
+
+    let (mut local_server_handle, _, mut upstream_server_handle, _) =
+        common::setup_brokers(AllowAll, AllowAll);
+
+    let mut local_client = TestClientBuilder::new(local_server_handle.address())
+        .with_client_id(ClientId::IdWithExistingSession("local_client".into()))
+        .build();
+
+    let mut upstream_client = TestClientBuilder::new(upstream_server_handle.address())
+        .with_client_id(ClientId::IdWithExistingSession("upstream_client".into()))
+        .build();
+
+    let dir = tempfile::tempdir().expect("Failed to create temp dir");
+    let storage_dir_override = dir.path().to_path_buf();
+
+    let (controller_handle, controller_task) = common::setup_bridge_controller(
+        local_server_handle.address(),
+        upstream_server_handle.tls_address().unwrap(),
+        subs.clone(),
+        &storage_dir_override,
+    )
+    .await;
+
+    local_client
+        .subscribe("downstream/filter/#", QoS::AtLeastOnce)
+        .await;
+
+    // wait to receive subscription ack
+    local_client.subscriptions().next().await;
+
+    upstream_client
+        .subscribe("upstream/temp/#", QoS::AtLeastOnce)
+        .await;
+
+    // wait to receive subscription ack
+    upstream_client.subscriptions().next().await;
+
+    // send upstream
+    local_client
+        .publish_qos1("to/temp/1", "from local", false)
+        .await;
+
+    assert_matches!(
+        upstream_client.publications().next().await,
+        Some(ReceivedPublication{payload, .. }) if payload == Bytes::from("from local")
+    );
+
+    // send upstream but disconnect afterwards
+    local_client
+        .publish_qos1("to/temp/1", "from local again", false)
+        .await;
+
+    controller_handle.shutdown();
```

Comment:
I believe there is no guarantee that this test does what you want. There is no way to prove that the message has not been already sent to upstream before shutdown. Maybe what you should do is this:
- shutdown upstream_server_handle and wait for upstream broker task - - it will guarantee that the upstream bridge loses connection to the upstream broker.
- publish a retain message - it will get to the bridge storage.
- [optional] - if you want to verify local restart - you can restart bridge controller at this step.
- restart the upstream broker.
- re-connect and re-subscribe the upstream client. - since message was "retained", it is guaranteed that re-connected client will receive it even if it re-subscribed with delay.
- assert that message received.